perception/autoware_map_based_prediction/src/predictor_vru.cpp

```diff
@@ -257,6 +260,64 @@ bool PredictorVru::doesPathCrossAnyFenceBeforeCrosswalk(
   return false;
 }
 
+PredictedPath PredictorVru::cutPathBeforeFences(const PredictedPath & predicted_path) const
+{
+  lanelet::BasicLineString2d predicted_path_ls;
+  for (const auto & pt : predicted_path.path) {
+    predicted_path_ls.emplace_back(pt.position.x, pt.position.y);
+  }
+  const auto candidates =
+    fence_layer_->lineStringLayer.search(lanelet::geometry::boundingBox2d(predicted_path_ls));
+  std::vector<lanelet::ConstLineString3d> crossed_fences{};
+  for (const auto & candidate : candidates) {
+    if (doesPathCrossFence(predicted_path_ls, candidate)) {
+      crossed_fences.push_back(candidate);
+      break;
+    }
+  }
+  if (crossed_fences.empty()) {
+    return predicted_path;
+  }
+
+  const auto & path = predicted_path.path;
+  auto closest_cross_index = path.size() - 1;
+  double dist_to_closest_cross = std::numeric_limits<double>::infinity();
+  const double z = path.at(0).position.z;
+  for (const auto & fence : crossed_fences) {
+    std::vector<autoware_utils_geometry::Point2d> intersections;
+    boost::geometry::intersection(
+      predicted_path_ls, lanelet::utils::to2D(fence).basicLineString(), intersections);
+    if (intersections.empty()) {
+      continue;
+    }
+    unsigned nearest_intersection_index = path.size() - 1;
+    for (const auto & intersection : intersections) {
+      const auto nearest_index_this = autoware::motion_utils::findNearestIndex(
+        predicted_path.path,
+        autoware_utils_geometry::create_point(intersection.x(), intersection.y(), z));
+      if (nearest_index_this < nearest_intersection_index) {
+        nearest_intersection_index = nearest_index_this;
+      }
+    }
+    const double distance_to_nearest_intersection = std::inner_product(
+      path.begin(), path.begin() + nearest_intersection_index, path.begin() + 1, 0.0 /* init */,
+      std::plus<double>{}, [](const auto & a, const auto & b) {
+        return autoware_utils_geometry::calc_distance3d(a, b);
+      });
+    if (distance_to_nearest_intersection < dist_to_closest_cross) {
+      dist_to_closest_cross = distance_to_nearest_intersection;
+      closest_cross_index = nearest_intersection_index;
+    }
+  }
```

Comment:
I think the search for the `closest_cross_index` can be simplified.
```suggestion
  std::optional<size_t> closest_cross_index;
  for (auto i = 0UL; i + 1 < predicted_path_ls.size() && !closest_cross_index.has_value(); ++i) {
    autoware_utils_geometry::Segment2d path_segment(predicted_path_ls[i], predicted_path_ls[i+1]);
    for (const auto & fence : crossed_fences) {
      if(boost::geometry::intersects(path_segment, lanelet::utils::to2D(fence).basicLineString())) {
        closest_cross_index = i;
      }
    }
  }
```